replica-common/src/main/java/com/graphhopper/customspeeds/CustomSpeedsUtils.java

```diff
@@ -0,0 +1,81 @@
+package com.graphhopper.customspeeds;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.MoreCollectors;
+import com.graphhopper.config.Profile;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.Reader;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class CustomSpeedsUtils {
+
+    public static final String CUSTOM_SPEED_FILE_CONFIG_FIELD = "custom_speed_file";
+    private static final String CUSTOM_SPEED_OSM_WAY_ID_COL_NAME = "osm_way_id";
+    private static final String CUSTOM_SPEED_MAX_SPEED_KPH_COL_NAME = "max_speed_kph";
+
+    /**
+     * @param profiles list of profiles for which the custom speed mapping should be retrieved
+     * @return map of vehicle name to mapping from OSM way id to the custom speed to use for that way, in kph. vehicles
+     * without custom speeds are omitted from the map.
+     * @throws RuntimeException if any custom speed files could not be found or are not properly formatted
+     */
+    public static ImmutableMap<String, ImmutableMap<Long, Double>> getVehicleNameToCustomSpeeds(List<Profile> profiles) {
+        Map<String, File> vehicleNameToCustomSpeedFile = getVehicleNameToCustomSpeedFile(profiles);
+        return vehicleNameToCustomSpeedFile.entrySet()
+                .stream()
+                .collect(ImmutableMap.toImmutableMap(
+                        Map.Entry::getKey,
+                        entry -> CustomSpeedsUtils.parseOsmWayIdToMaxSpeed(entry.getValue())));
+    }
+
+    private static Map<String, File> getVehicleNameToCustomSpeedFile(List<Profile> profiles) {
+        Map<String, File> vehicleNameToCustomSpeedFileName = new HashMap<>();
+
+        for (Profile profile : profiles) {
+            String customSpeedFilePath = profile.getHints().getString(CustomSpeedsUtils.CUSTOM_SPEED_FILE_CONFIG_FIELD, "");
+            if (customSpeedFilePath.isEmpty()) {
+                continue;
+            }
+            File customSpeedFile = new File(customSpeedFilePath);
+
+            String vehicle = profile.getVehicle();
```

Comment:
The one big-ish question I have is about this line - are we sure we want `.getVehicle()`, and not `getName()`? Do these correspond exactly with `vehicle` and `name` laid out in our graphhopper configs? If so, I'm concerned that many configs have identical vehicle fields (eg `car`), while what I think we want is to associate specific profiles (eg `custom_car_fast_thurton_drive`) with specific speed CSVs.

Or is `getVehicle()` actually doing that?

Reply:
Good question. The `ReplicaVehicleTagParserFactory` (and the more general `VehicleTagParserFactory`) is able to customize the speeds for each way on a per-vehicle basis instead of a per-profile one. The `name` argument in the `VehicleTagParser createParser(EncodedValueLookup lookup, String name, PMap configuration)` method is the name of the vehicle to customize. I don't think GraphHopper offers any way of customizing the speeds for a given profile on a per-way basis; custom models can do some limited speed customization, but they don't get per-way granularity.

So therefore, every custom speed file must be associated with a unique vehicle, and internally we're actually applying those custom speeds to the vehicle rather than the profile. The uniqueness check is in this method right below. Will be sure to cover this requirement in the docs, but I don't think it should limit us. I used the `custom_car_fast_thurton_drive` name for both the profile and the vehicle and will recommend others do the same.

Reply:
Ahhh I see, I think the one detail here I glossed over is that you also set the vehicle name to something distinct in the test config (`custom_car_fast_thurton_drive`) - I'd read it at first as though you'd used `car` for that profile as well. Good to know about the background reasoning for this, as long as unique names are used for vehicles where custom speeds are being provided I can't see an issue.

> maybe we should prevent people from associating a vehicle with a custom speed file in one profile, then reusing that same vehicle name in a different profile that doesn't have a custom speed file? 

This sounds like a great idea to me! I think enforcing uniqueness can help us avoid sticky situations if users aren't aware of some of these subtleties

Reply:
cool, added in the stricter enforcement i mentioned above. good catch, thanks for asking the question!

also created `CustomSpeedUtilsTest` which tests some details not covered by `RouterServerTest` - parsing multiline custom speed files, using multiple profiles with custom speeds, and the unique vehicle+custom speed file pairing requirement

Reply:
Awesome! Took me a bit to chew over how you used `Optional` in here, but it makes sense once I wrapped my head around it again, and the tests confirm the behavior well (thanks a bunch for adding those!). It's good we're thinking through all these combinations of use cases, things get tricky fast 🙃